ovlfs-2.0.1/doc/ovlfstab.5
.\"
.\" COPYRIGHT (C) 1998-2003 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: ovlfstab.m
.\"
.\" DESCRIPTION:
.\"	Manual page for the ovlfstab mount tab file's format.
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 2003-06-01	ARTHUR N.	Initial Release
.\"
.\"
.\" M4 definitions:
.\" COPYRIGHT (C) 1998 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: tmac.asn
.\"
.\" DESCRIPTION: This file contains general purpose nroff/troff macros.
.\"
.\" SCCS ID: @(#) tmac.asn 1.1
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 03/08/1998	ARTHUR N.	Initial Release
.\"
.\"
.\" **************************************************************************
.\"
.\" START OF LIST MACROS
.\"
.de BL    \" Begin Bullet List
.in +8n             \" increase indent by about 8 characters
.nr *i \\n(.i/1n    \" remember the current indent in units of 'en'
.nr *i +2           \" add two more characters to the list item indent
.ds OL \\*(LM       \" remember the last mark - can only nest 2 levels
.ds LM \(bu         \" set the list item marker to a bullet
..
.de LE    \" End List
.nr *i -10      \" calculate original indent
.in \\n(*in     \" reset original indent
.ds LM \\*(OL   \" restore last marker saved
..
.de LI    \" List Item
.in \\n(*in    \" indent to list item space
.br            \" make sure to start a new line
.ti -2n        \" indent, next line only, back about 2 characters
\\*(LM         \" display list marker
..
.\"
.\" **************************************************************************
.\"
.\" START OF THE HANGING INDENT MACRO
.\"
.\" begin hanging indent from current point after the text of argument 1,
.\"  if defined.
.\"
.nr %I 0
.nr w 1n	\" Set this once and allow it to be changed later
.de HI
.if \\n(%I-9  .ab HI macro stack error    \" If %I > 9, can't continue
.nr I\\n(%I \\n(.iu
.nr %I \\n(%I+1
.di *H           \" Divert output to *H macro and put $1 on input
\&\\$1
.br              \" Break this input to put in output (so the diversion gets it)
.di              \" End the diversion
.nr $i \\n(dlu+\\nwu  \" Calculate the hanging indentation
.nr %i \\nw-\w'  '    \" Take offset and subtract the spaces that are coming
.ti 0
\&\\*(*H\\h'\\n(%iu'
'in \\n($iu
..
.\"
.\" end of hanging indent - restore the PREVIOUS indent
.\"
.de HE
.nr %I \\n(%I-1       \" Decrement current indentation counter
.in \\n(I\\n(%Iu      \" Restore indentation at current counter
..
.\"
.\" END OF THE HANGING INDENT MACRO
.\"
.\" **************************************************************************
.\"
.\" START OF THE TAB FIELD MACRO
.\"
.\" The .Tf macro moves over to the specified horrizontal position on the line
.\"  If the current input is past the requested position, then a break is
.\"  inserted followed by the movement to the requested position.
.\"
.\" NOTE: the positioning does not work correctly for troff; I believe the
.\"       problem is the length of the diversion is different than the length
.\"       of the string when output.  The results are close, but not perfectly
.\"       aligned
.\"
.de Tf  \" tab field
.\"		Check argument
.if \\n(.$<1 .ab Tf macro requires an argument
.\"		Find the length of the current line: use a diversion to get it
.di }s
.br
.di
.\"		Find length of the diversion plus the trailing space that will
.\"		follow it
.di }t
'ti 0           \"	Indent already part of string }s
\\*(}s\c\ 
.br
.di
.\"		Calculate the length of the line (minus the current indent)
.nr }u \\n(dlu-\\n(.iu
.br
.\"		Now put the original input line back out followed by the
.\"		horizontal positioning requested
'ti 0
\\*(}s\c            \"		Again, indent is already part of string }s
.ie \\n(}u-\\$1>0 \{
.br
\\h'\\$1-1m'\c \}
.el \\h'\\$1-\\n(}uu-1m'\c
..
.\"
.\"
.\"
.ad l
.TH ovlfstab 5 "Sun Jul 27 22:21:37 EDT 2003" "Arthur Naseef" "Linux System Admin"
.\"
.\" SECTION BREAK
.\"
.SH NAME
ovlfstab -
format of the
mount table
for ovlfsmount.
.\"
.\" SECTION BREAK
.\"
.SH VERSION
This manual page covers the version of the
ovlfsmount program which corresponds to
version 2.0.1 of the overlay filesystem.
.\"
.\" SECTION BREAK
.\"
.SH DESCRIPTION
Filesystem and selector entries
are defined
in the ovlfstab file
for use
by the \fBovlfsmount(8)\fR
program.
The format
of the file
is described below.
Please see
the manual page
for the \fBovlfsmount(8)\fR
program for more information.
.\"
.\" SUB-SECTION BREAK
.\"
.SS File Format
Below is a description
of the overall
file format.
Detailed descriptions
of each element
of the file format
are given later.
.\"
.P
The file consists
of one or more
of following elements:
.BL
.P
.LI
Prefilter Command
.P
.LI
Filesystem Declaration
.P
.LI
Filesystem Selector Declaration
.LE
.\"
.P
Each of these
may be used
multiple times,
in any order.
In addition
to these elements,
comments may be
placed on any line
in the file.
.\"
.\" SUB-SECTION BREAK
.\"
.SS Prefilter Command
Using the format below,
the prefilter command
changes the interpretation of
(i.e. filters)
the file
until the end of file
is reached.
Note that successive
filters work inline
with previous ones,
with the output
of earlier filters
providing input
to later filters.
The format
of a prefilter command
is as follows:
.sp 1
.ti +1i
#!! <command\-string>
.sp 1
where <command\-string>
is a string,
terminated by a newline,
interpreted by /bin/sh.
For example,
the following input:
.sp 1
.ti +1i
#!! /bin/m4 -DHOME=$HOME
.sp 1
would be interpreted
as follows:
.sp 1
.ti +1i
/bin/sh -c "/bin/m4 -DHOME=$HOME"
.\"
.\" SUB-SECTION BREAK
.\"
.SS Filesystem Declaration
Using the syntax below,
a single filesystem mount
can be specified
and given a name.
In addition to 
the mount point,
storage root,
and base root,
all filesystem options
can be specified.
.sp 1
.ti +1i
FILESYSTEM <name> <mount\-parms> END
.sp 1
where <name> is a string
that names the filesystem,
and <mount\-parms> defines
the mount point
and other mount parameters.
Please see the section,
\fBMOUNT PARAMETERS\fR,
below for a list
of parameters
and their syntax.
.\"
.P
The following
is an example declaration
of a filesystem
named "safe-root":
.sp 1
.nf
    FILESYSTEM "safe-root"
        mount-pt    /mnt/safe
        base_root   /
        storage     /ovl/safe-root
        options
          "magic,hidemagic,smagic=Magic_stg,bmagic=Magic_base"
    END
.fi
.\"
.\" SUB-SECTION BREAK
.\"
.SS Filesystem Selector Declaration
Two types of selectors exist,
CD and special.
The syntax of both
types of selector statements
is as follows:
.sp 1
.in +1i
.HI "CD SELECTOR"
<name> [<selector\-table>]
.HE
.br
END
.sp 1
.HI "SPECIAL SELECTOR"
<name> <selector\-command> [<selector\-table>]
.HE
.br
END
.\"
.sp 1
where <name>
is a string
that identifies the selector,
<selector\-command> is a string
used to determine the selector key
at mount time,
and <selector\-table>
maps keys to filesystem names,
defining which filesystem
is mounted for each key.
.\"
.P
The syntax for <selector\-command>
is as follows:
.sp 1
.ti +1i
COMMAND <program> [<arg> ...]
.sp 1
where <program> is a string
that specifies the
program to execute,
and <arg> is a string
passed to the program
as an argument.
For example:
.sp 1
.ti +1i
COMMAND "/usr/local/cd_vol_id" "-d" "/dev/cdrom"
.sp 1
Mapping of the selector keys
to filesystem names,
and, optionally, additional options
for each mount,
may be specified
in the <selector\-table>.
The table may contain
any number of entries;
the format of each entry
is as follows:
.sp 1
.ti +1i
KEY <key> <fs\-name> [<mount\-parm> ...]
.sp 1
where <key> is a string
containing the selector key,
<fs\-name> is a string
containing the name
of the filesystem
to be mounted,
and <mount\-parm> contains
a mount parameter.
Mount parameters specified
in the table
take precedence over
the same options
in the filesystem declaration.
.\"
.\" SUB-SECTION BREAK
.\"
.SS Comments
Limited commenting
is available
in the ovlfstab file.
Every line that
begins with the character,
\fB#\fR,
is considered a comment line.
All content through
the next newline is ignored.
Note that the \fB#\fR may
be preceeded by whitespace.
.\"
.\" SECTION BREAK
.\"
.SH MOUNT PARAMETERS
Below is a description
of each of the mount parameters
supported by \fBovlfsmount(8)\fR.
The syntax of each
is the parameter name
immediately followed
by a string
containing the value
for the parameter.
.TP
\fBBASE_ROOT\fR
Path to the root directory
for the mount's base filesystem.
.TP
\fBOPTIONS\fR
Mount options
passed to the ovlfs
at mount time.
.TP
\fBMOUNT-POINT\fR or \fBMOUNT-PT\fR
Mount point for the filesystem.
Access to the ovlfs
will be rooted
at this point.
.TP
\fBSTG_FILE\fR
Path to the storage file
into which the ovlfs
inodes and other
filesystem information
is stored persistently.
.TP
\fBSTG_METHOD\fR
Name of the storage method
used by the mount.
As of this writing,
only the storage method
named "Storage File"
is available.
.TP
\fBSTORAGE\fR
Path to the root directory
for the mount's storage filesystem.
.\"
.\" SECTION BREAK
.\"
.SH STRING FORMATS
Strings in the file
can be given
in one of
the following formats:
.BL
.sp 1
.LI
.HI "Quoted -"
Started with a
double or single quote,
all text
up to the next
unescaped, matching quote,
including whitespace
is included in the string.
The backslash character
is used to escape
the immediately following character.
No other special handling
is performed.
.sp 1
.LI
.HI "Special -"
Preceeded by the
total length of
the string,
enclosed in braces,
the string is
otherwise read from
the file as-is.
For example,
{4}/mnt is equivalent
to "/mnt".
.sp 1
.LI
.HI "Normal -"
Any string of
non-whitespace
characters in the file,
that is not a keyword,
is interpreted
as a string.
.\"
.P
Also note that adjacent strings,
without any whitespace
separating them,
are concatenated into
a single string.
.\"
.\" SECTION BREAK
.\"
.SH KEYWORDS
The following keywords
are recognized
by the \fBovlfsmount(8)\fR program
as of this writing.
Note that keywords
are case-insensitive.
.sp 1
.BL
.LI
BASE_ROOT
.LI
CD
.LI
COMMAND
.LI
END
.LI
FILESYSTEM
.LI
KEY
.LI
OPTIONS
.LI
MOUNT-POINT
.LI
MOUNT-PT
.LI
SELECTOR
.LI
SPECIAL
.LI
STG_FILE
.LI
STG_METHOD
.LI
STORAGE
.LE
.\"
.\" SECTION BREAK
.\"
.SH EXAMPLE
.nf
# safe-root creates an environment that can be used with chroot to give full
#  access to the root filesystem while protecting it from modification.

FILESYSTEM "safe-root"
	mount-pt     "/ovlfs/safe-root"
	base_root    "/"
	storage      "/ovlfs/safe-root.stg"
	stg_file     "/ovlfs_stg/safe-root.dat"
	options
          "magic,hidemagic,smagic=Magic_stg,bmagic=Magic_base"
END


# cd-video01 mounts a CD named "Video 01" (note the CD must already be mounted
#  on /mnt/cdrom).

FILESYSTEM "cd-video01"
	mount-pt    "/ovlfs/cd"
	base_root   "/mnt/cdrom"
	storage     "/ovlfs/cd-video01"
	stg_file    "/ovlfs_stg/cd-video01.dat"
END


# cd-video02 mounts a CD named "Video 02" (note the CD must already be mounted
#  on /mnt/cdrom).

FILESYSTEM "cd-video02"
	mount-pt    "/ovlfs/cd"
	base_root   "/mnt/cdrom"
	storage     "/ovlfs/cd-video02"
	stg_file    "/ovlfs_stg/cd-video02.dat"
END


# Select cd-video01 or cd-video02.

CD SELECTOR "cd-video"
	KEY "10A9FE3000AA5903"	"cd-video01"
	KEY "3A602AAB7FB4776E"	"cd-video02" options "nomagic"
END
.fi
.\"
.\" SECTION BREAK
.\"
.SH EXAMPLE WITH PREFILTERING
.nf
#!!/usr/bin/m4 -DOVL_MNTS_DIR=/ovlfs -DOVL_STGFILE_DIR=/ovlfs_stg
define(m4_concat, $1$2$3$4$5$6$7$8$9)dnl

# safe-root creates an environment that can be used with chroot to give full
#  access to the root filesystem while protecting it from modification.

FILESYSTEM "safe-root"
	mount-pt     m4_concat(OVL_MNTS_DIR, "/safe-root")
	base_root    "/"
	storage      m4_concat(OVL_MNTS_DIR, "/safe-root.stg")
	stg_file     m4_concat(OVL_STGFILE_DIR, "/safe-root.dat")
	options
          "magic,hidemagic,smagic=Magic_stg,bmagic=Magic_base"
END

# cd-video01 mounts a CD named "Video 01" (note the CD must already be mounted
#  on /mnt/cdrom).

FILESYSTEM "cd-video01"
	mount-pt    m4_concat(OVL_MNTS_DIR, "/cd")
	base_root   "/mnt/cdrom"
	storage     m4_concat(OVL_MNTS_DIR, "/cd-video01")
	stg_file    m4_concat(OVL_STGFILE_DIR, "/cd-video01.dat")
END


# cd-video02 mounts a CD named "Video 02" (note the CD must already be mounted
#  on /mnt/cdrom).

FILESYSTEM "cd-video02"
	mount-pt    m4_concat(OVL_MNTS_DIR, "/cd")
	base_root   "/mnt/cdrom"
	storage     m4_concat(OVL_MNTS_DIR, "/cd-video02")
	stg_file    m4_concat(OVL_STGFILE_DIR, "/cd-video02.dat")
END


# Select cd-video01 or cd-video02.

CD SELECTOR "cd-video"
	KEY "10A9FE3000AA5903"	"cd-video01"
	KEY "3A602AAB7FB4776E"	"cd-video02" options "nomagic"
END
.fi
.\"
.\" SECTION BREAK
.\"
.SH SEE ALSO
ovl(8),
ovlfsmount(8)
.\"
.\" SECTION BREAK
.\"
.SH COPYRIGHT
This manual page and associated source code is licensed under the
GNU General Public License Version 2.
Of course, all source derived from the linux kernel source is copyrighted
by the original author of that source.
.sp 1
COPYRIGHT (C) 2003 By Arthur Naseef
.\"
.\" SECTION BREAK
.\"
.SH FILE VERSION
 @(#)ovlfstab.m	1.1	[03/07/27 22:20:33]


ovlfs-2.0.1/RELEASE_NOTES
                               ==================
                               OVERLAY FILESYSTEM
                                 RELEASE NOTES
                                 VERSION 2.0.0
                               ==================


1. ROBUSTNESS
=============

    Due to the amount of time passed since the initially planned release of
the 2.0.0 version, the following has been affected:

	A. Some of the documentation is not up-to-date.

	B. New files and code have been added with no real impact on the
	   software.

	C. Testing with the latest version of the 2.4 kernel has not been
	   completed.

	D. New features are not documented.

	E. Old warnings and issues may not be applicable.

    In spite of these, this software is working and stable enough for release.
I hope you will find this to be the case.

    Note that support for the ReiserFS as the base or storage filesystem is
not complete; multiple hard links to files, and renamed files in the base or
storage filesystem may not be properly resolved by the Overlay Filesystem.
Otherwise, ReiserFS support is functional.



2. KERNEL VERSIONS
==================

    This version of the Overlay Filesystem is built on, and supports, 2.4
series kernels.  Please note that a limited set of 2.4 kernels have actually
been used, so there may be incompatibilities with others due to changes in the
VFS or other portions of the Linux kernel.

    While support for 2.2 kernels was considered, it is not complete; the 2.2
and 2.4 kernels are very similar, but 2.2 lacks many useful interfaces on which
the Overlay Filesystem relies.  Therefore, 2.2 support is currently incomplete.

    Support for 2.0 kernels was present in the 1.0 series for the Overlay
Filesystem, but that support has been lost in the migration to 2.4; in
addition, the 1.0 ovlfs versions had serious flaws that have been corrected.



3. BUG FIXES
============

    Since 2.0.0 is a major rewrite from version 1, listing bug fixes is not
effective.  However, there were serious flaws in version 1 that have been
corrected.  These are listed here:

	A. Directory hierarchy creation used a recursive method that could
	   easily overflow the kernel stack.

	B. Large stack variables were allocated by the ovlfs, which could
	   easily overflow the kernel stack.  All such variables have been
	   replaced with heap memory.

	C. A storage filesystem root that could be reached under the base
	   filesystem root was not supported, and such an instance lead to
	   certain, serious problems in the kernel.  This is no longer true.



4. IMPROVEMENTS
===============

    Many improvements have been made to the Overlay Filesystem.  The majority
support future development efforts and improve the stability and operation of
the filesystem.

    The most notable improvement is the Storage System support which allows
alternative methods of storing Overlay Filesystem data to be added to the core
without changing the core.  Support for storing the filesystem's data on block
devices and in a block-formatted file are planned, which will have a great
impact on the filesystem's required memory use.

    Another notable improvement is the ovlmount program included in this
release.  It simplifies the management of multiple Overlay Filesystem mounts
by using an fstab-like file specifically designed for the Overlay Filesystem.



5. SMP SUPPORT
==============

    At this time, SMP support is not tested, and is expected NOT to work.
The effort needed to find and resolve SMP issues has not been made yet.



6. DEVELOPER'S NOTE
===================

    Much remains to be cleaned up for "correct" kernel coding.  This holds
especially true in the case of module versions.  The current plan for multiple
kernel version support involves splitting the core into different directories,
with one for each major implementation based on the kernel version.  With this
model, the existing code can be greatly reduced in size and readability.

    Any help in this matter is greatly appreciated.



7. LIVE CD CREATED
==================

    With this version of the Overlay Filesystem, I have successfully created a
"live", bootable CD using the Overlay Filesystem as the root filesystem.  Note
that the real root is replaced after kernel startup using a small application
that I wrote.

    The CD boots directly to a SuSE 7.3 system with web browser and all types
of wonderful features.  If you would like to know more, please write!

    One of the great benefits of the live CD is the thorough testing it gives
the Overlay Filesystem!



8. LEX AND YACC
===============

    The mount directory uses lex and yacc (or flex and bison) to generate a
parser for the ovlfstab file.  Output from these programs is included in the
release by default, and will not be removed by "make clean", so that lex and
yacc should not be required on your system.  Note, though, that the flex
library is needed.

    If you prefer to rebuild the lex and yacc output, please note that the
makefiles expect programs named lex and yacc by default.



9. THANKS
=========

    Many thanks goes out to the User Mode Linux project hosted on Sourceforge
at the following URL:

	http://user-mode-linux.sourceforge.net/

Debugging the Overlay Filesystem has been made significantly easier, and
corruption resulting from bugs was limited to a "safe area" (i.e. the content
of a single file).

    Keep up the good work!



[version @(#) RELEASE_NOTES 1.3]


ovlfs-2.0.1/doc/ovlfstab.m
.\"
.\" COPYRIGHT (C) 1998-2003 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: ovlfstab.m
.\"
.\" DESCRIPTION:
.\"	Manual page for the ovlfstab mount tab file's format.
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 2003-06-01	ARTHUR N.	Initial Release
.\"
.\"
.\" M4 definitions:
include(tmac.asn)dnl
define(m4_concat, $1$2$3$4$5$6$7$8$9)dnl
define(CURRENT_DATE, translit(esyscmd(date), `
', `'))dnl
define(OVL_VERSION, translit(esyscmd(awk '{print $1; exit}' Version), `
', `'))dnl
.\"
.\"
.\"
.ad l
.TH ovlfstab 5 "CURRENT_DATE" "Arthur Naseef" "Linux System Admin"
.\"
.\" SECTION BREAK
.\"
.SH NAME
ovlfstab -
format of the
mount table
for ovlfsmount.
.\"
.\" SECTION BREAK
.\"
.SH VERSION
This manual page covers the version of the
ovlfsmount program which corresponds to
version OVL_VERSION of the overlay filesystem.
.\"
.\" SECTION BREAK
.\"
.SH DESCRIPTION
Filesystem and selector entries
are defined
in the ovlfstab file
for use
by the \fBovlfsmount(8)\fR
program.
The format
of the file
is described below.
Please see
the manual page
for the \fBovlfsmount(8)\fR
program for more information.
.\"
.\" SUB-SECTION BREAK
.\"
.SS File Format
Below is a description
of the overall
file format.
Detailed descriptions
of each element
of the file format
are given later.
.\"
.P
The file consists
of one or more
of following elements:
.BL
.P
.LI
Prefilter Command
.P
.LI
Filesystem Declaration
.P
.LI
Filesystem Selector Declaration
.LE
.\"
.P
Each of these
may be used
multiple times,
in any order.
In addition
to these elements,
comments may be
placed on any line
in the file.
.\"
.\" SUB-SECTION BREAK
.\"
.SS Prefilter Command
Using the format below,
the prefilter command
changes the interpretation of
(i.e. filters)
the file
until the end of file
is reached.
Note that successive
filters work inline
with previous ones,
with the output
of earlier filters
providing input
to later filters.
The format
of a prefilter command
is as follows:
.sp 1
.ti +1i
`#!!' <command\-string>
.sp 1
where <command\-string>
is a string,
terminated by a newline,
interpreted by /bin/sh.
For example,
the following input:
.sp 1
.ti +1i
`#!!' /bin/m4 -DHOME=$HOME
.sp 1
would be interpreted
as follows:
.sp 1
.ti +1i
`/bin/sh -c "/bin/m4 -DHOME=$HOME"'
.\"
.\" SUB-SECTION BREAK
.\"
.SS Filesystem Declaration
Using the syntax below,
a single filesystem mount
can be specified
and given a name.
In addition to 
the mount point,
storage root,
and base root,
all filesystem options
can be specified.
.sp 1
.ti +1i
FILESYSTEM <name> <mount\-parms> END
.sp 1
where <name> is a string
that names the filesystem,
and <mount\-parms> defines
the mount point
and other mount parameters.
Please see the section,
\fBMOUNT PARAMETERS\fR,
below for a list
of parameters
and their syntax.
.\"
.P
The following
is an example declaration
of a filesystem
named "safe-root":
.sp 1
.nf
    FILESYSTEM "safe-root"
        mount-pt    /mnt/safe
        base_root   /
        storage     /ovl/safe-root
        options
          "magic,hidemagic,smagic=Magic_stg,bmagic=Magic_base"
    END
.fi
.\"
.\" SUB-SECTION BREAK
.\"
.SS Filesystem Selector Declaration
Two types of selectors exist,
CD and special.
The syntax of both
types of selector statements
is as follows:
.sp 1
.in +1i
.HI "CD SELECTOR"
<name> [<selector\-table>]
.HE
.br
END
.sp 1
.HI "SPECIAL SELECTOR"
<name> <selector\-command> [<selector\-table>]
.HE
.br
END
.\"
.sp 1
where <name>
is a string
that identifies the selector,
<selector\-command> is a string
used to determine the selector key
at mount time,
and <selector\-table>
maps keys to filesystem names,
defining which filesystem
is mounted for each key.
.\"
.P
The syntax for <selector\-command>
is as follows:
.sp 1
.ti +1i
COMMAND <program> [<arg> ...]
.sp 1
where <program> is a string
that specifies the
program to execute,
and <arg> is a string
passed to the program
as an argument.
For example:
.sp 1
.ti +1i
COMMAND "/usr/local/cd_vol_id" "-d" "/dev/cdrom"
.sp 1
Mapping of the selector keys
to filesystem names,
and, optionally, additional options
for each mount,
may be specified
in the <selector\-table>.
The table may contain
any number of entries;
the format of each entry
is as follows:
.sp 1
.ti +1i
KEY <key> <fs\-name> [<mount\-parm> ...]
.sp 1
where <key> is a string
containing the selector key,
<fs\-name> is a string
containing the name
of the filesystem
to be mounted,
and <mount\-parm> contains
a mount parameter.
Mount parameters specified
in the table
take precedence over
the same options
in the filesystem declaration.
.\"
.\" SUB-SECTION BREAK
.\"
.SS Comments
Limited commenting
is available
in the ovlfstab file.
Every line that
begins with the character,
\fB#\fR,
is considered a comment line.
All content through
the next newline is ignored.
Note that the \fB#\fR may
be preceeded by whitespace.
.\"
.\" SECTION BREAK
.\"
.SH MOUNT PARAMETERS
Below is a description
of each of the mount parameters
supported by \fBovlfsmount(8)\fR.
The syntax of each
is the parameter name
immediately followed
by a string
containing the value
for the parameter.
.TP
\fBBASE_ROOT\fR
Path to the root directory
for the mount's base filesystem.
.TP
\fBOPTIONS\fR
Mount options
passed to the ovlfs
at mount time.
.TP
\fBMOUNT-POINT\fR or \fBMOUNT-PT\fR
Mount point for the filesystem.
Access to the ovlfs
will be rooted
at this point.
.TP
\fBSTG_FILE\fR
Path to the storage file
into which the ovlfs
inodes and other
filesystem information
is stored persistently.
.TP
\fBSTG_METHOD\fR
Name of the storage method
used by the mount.
As of this writing,
only the storage method
named "Storage File"
is available.
.TP
\fBSTORAGE\fR
Path to the root directory
for the mount's storage filesystem.
.\"
.\" SECTION BREAK
.\"
.SH STRING FORMATS
Strings in the file
can be given
in one of
the following formats:
.BL
.sp 1
.LI
.HI "Quoted -"
Started with a
double or single quote,
all text
up to the next
unescaped, matching quote,
including whitespace
is included in the string.
The backslash character
is used to escape
the immediately following character.
No other special handling
is performed.
.sp 1
.LI
.HI "Special -"
Preceeded by the
total length of
the string,
enclosed in braces,
the string is
otherwise read from
the file as-is.
For example,
`{4}/mnt' is equivalent
to "/mnt".
.sp 1
.LI
.HI "Normal -"
Any string of
non-whitespace
characters in the file,
that is not a keyword,
is interpreted
as a string.
.\"
.P
Also note that adjacent strings,
without any whitespace
separating them,
are concatenated into
a single string.
.\"
.\" SECTION BREAK
.\"
.SH KEYWORDS
The following keywords
are recognized
by the \fBovlfsmount(8)\fR program
as of this writing.
Note that keywords
are case-insensitive.
.sp 1
.BL
.LI
BASE_ROOT
.LI
CD
.LI
COMMAND
.LI
END
.LI
FILESYSTEM
.LI
KEY
.LI
OPTIONS
.LI
MOUNT-POINT
.LI
MOUNT-PT
.LI
SELECTOR
.LI
SPECIAL
.LI
STG_FILE
.LI
STG_METHOD
.LI
STORAGE
.LE
.\"
.\" SECTION BREAK
.\"
.SH EXAMPLE
.nf
# safe-root creates an environment that can be used with chroot to give full
#  access to the root filesystem while protecting it from modification.

FILESYSTEM "safe-root"
	mount-pt     "/ovlfs/safe-root"
	base_root    "/"
	storage      "/ovlfs/safe-root.stg"
	stg_file     "/ovlfs_stg/safe-root.dat"
	options
          "magic,hidemagic,smagic=Magic_stg,bmagic=Magic_base"
END


# cd-video01 mounts a CD named "Video 01" (note the CD must already be mounted
#  on /mnt/cdrom).

FILESYSTEM "cd-video01"
	mount-pt    "/ovlfs/cd"
	base_root   "/mnt/cdrom"
	storage     "/ovlfs/cd-video01"
	stg_file    "/ovlfs_stg/cd-video01.dat"
END


# cd-video02 mounts a CD named "Video 02" (note the CD must already be mounted
#  on /mnt/cdrom).

FILESYSTEM "cd-video02"
	mount-pt    "/ovlfs/cd"
	base_root   "/mnt/cdrom"
	storage     "/ovlfs/cd-video02"
	stg_file    "/ovlfs_stg/cd-video02.dat"
END


# Select cd-video01 or cd-video02.

CD SELECTOR "cd-video"
	KEY "10A9FE3000AA5903"	"cd-video01"
	KEY "3A602AAB7FB4776E"	"cd-video02" options "nomagic"
END
.fi
.\"
.\" SECTION BREAK
.\"
.SH EXAMPLE WITH PREFILTERING
.nf
#!!/usr/bin/m4 -DOVL_MNTS_DIR=/ovlfs -DOVL_STGFILE_DIR=/ovlfs_stg
`define(m4_concat, $1$2$3$4$5$6$7$8$9)dnl'

# safe-root creates an environment that can be used with chroot to give full
#  access to the root filesystem while protecting it from modification.

FILESYSTEM "safe-root"
	mount-pt     `m4_concat'(OVL_MNTS_DIR, "/safe-root")
	base_root    "/"
	storage      `m4_concat'(OVL_MNTS_DIR, "/safe-root.stg")
	stg_file     `m4_concat'(OVL_STGFILE_DIR, "/safe-root.dat")
	options
          "magic,hidemagic,smagic=Magic_stg,bmagic=Magic_base"
END

# cd-video01 mounts a CD named "Video 01" (note the CD must already be mounted
#  on /mnt/cdrom).

FILESYSTEM "cd-video01"
	mount-pt    `m4_concat'(OVL_MNTS_DIR, "/cd")
	base_root   "/mnt/cdrom"
	storage     `m4_concat'(OVL_MNTS_DIR, "/cd-video01")
	stg_file    `m4_concat'(OVL_STGFILE_DIR, "/cd-video01.dat")
END


# cd-video02 mounts a CD named "Video 02" (note the CD must already be mounted
#  on /mnt/cdrom).

FILESYSTEM "cd-video02"
	mount-pt    `m4_concat'(OVL_MNTS_DIR, "/cd")
	base_root   "/mnt/cdrom"
	storage     `m4_concat'(OVL_MNTS_DIR, "/cd-video02")
	stg_file    `m4_concat'(OVL_STGFILE_DIR, "/cd-video02.dat")
END


# Select cd-video01 or cd-video02.

CD SELECTOR "cd-video"
	KEY "10A9FE3000AA5903"	"cd-video01"
	KEY "3A602AAB7FB4776E"	"cd-video02" options "nomagic"
END
.fi
.\"
.\" SECTION BREAK
.\"
.SH SEE ALSO
ovl(8),
ovlfsmount(8)
.\"
.\" SECTION BREAK
.\"
.SH COPYRIGHT
This manual page and associated source code is licensed under the
GNU General Public License Version 2.
Of course, all source derived from the linux kernel source is copyrighted
by the original author of that source.
.sp 1
COPYRIGHT (C) 2003 By Arthur Naseef
.\"
.\" SECTION BREAK
.\"
.SH FILE VERSION
 @(#)ovlfstab.m	1.1	[03/07/27 22:20:33]
